From 5984c95b94057c3b5b072158525cac6d5feef175 Mon Sep 17 00:00:00 2001
From: Stuart MENEFY <stuart.menefy@st.com>
Date: Tue, 29 Sep 2009 18:36:38 +0100
Subject: [PATCH] sh: Move the alignment exceptions flags

Commit f775a6d40e4812ce9f97af7b589cb8e26c3275e3 "sh: implement
PR_[GS]ET_UNALIGN prctls." introduced three new task flags (TIF_...)
which are used to record the misaligned handling behaviour requested
by the user through the prctl command.

However the flags were positioned in the lower byte of the flags word,
which means that they are checked when returning to user mode against
the work flags (_TIF_WORK_MASK and _TIF_ALLWORK_MASK), and they
shouldn't be. We could just change the masks, but as later kernel
versions use these bit positions for things which do have to be
included in the work masks, and the assembler code can be more
efficient if the work masks are 8 bits or less, move the misaligned
flags to the top half of the word.

Also drop the NOFIX flag as it was never used.

Signed-off-by: Stuart Menefy <stuart.menefy@st.com>
---
 include/asm-sh/thread_info.h |   12 +++++-------
 1 files changed, 5 insertions(+), 7 deletions(-)

diff --git a/include/asm-sh/thread_info.h b/include/asm-sh/thread_info.h
index ee0c483..2cf209b 100644
--- a/include/asm-sh/thread_info.h
+++ b/include/asm-sh/thread_info.h
@@ -112,13 +112,13 @@ static inline struct thread_info *current_thread_info(void)
 #define TIF_RESTORE_SIGMASK	3	/* restore signal mask in do_signal() */
 #define TIF_SINGLESTEP		4	/* singlestepping active */
 #define TIF_KERNEL_TRACE	5	/* kernel trace active */
-#define TIF_UAC_NOPRINT		6	/* Unaligned Access control flags ...*/
-#define TIF_UAC_SIGBUS		7
-#define TIF_UAC_NOFIX		8
 #define TIF_USEDFPU		16	/* FPU was used by this task this quantum (SMP) */
 #define TIF_POLLING_NRFLAG	17	/* true if poll_idle() is polling TIF_NEED_RESCHED */
 #define TIF_MEMDIE		18
 #define TIF_FREEZE		19
+#define TIF_UAC_NOPRINT		20	/* PR_UNALIGN_NOPRINT (1) */
+#define TIF_UAC_SIGBUS		21	/* PR_UNALIGN_SIGBUS (2) */
+
 
 #define _TIF_SYSCALL_TRACE	(1<<TIF_SYSCALL_TRACE)
 #define _TIF_SIGPENDING		(1<<TIF_SIGPENDING)
@@ -129,16 +129,14 @@ static inline struct thread_info *current_thread_info(void)
 #define _TIF_USEDFPU		(1<<TIF_USEDFPU)
 #define _TIF_POLLING_NRFLAG	(1<<TIF_POLLING_NRFLAG)
 #define _TIF_FREEZE		(1<<TIF_FREEZE)
-
-#define _TIF_UAC_SIGBUS		(1<<TIF_UAC_SIGBUS)
 #define _TIF_UAC_NOPRINT	(1<<TIF_UAC_NOPRINT)
-#define _TIF_UAC_NOFIX		(1<<TIF_UAC_NOFIX)
+#define _TIF_UAC_SIGBUS		(1<<TIF_UAC_SIGBUS)
 
 #define _TIF_WORK_MASK		0x000000DE	/* work to do on interrupt/exception return */
 #define _TIF_ALLWORK_MASK	0x000000FF	/* work to do on any return to u-space */
 
 /* PR_[GS]ET_UNALIGN prctls */
-#define SH_UAC_SHIFT		6
+#define SH_UAC_SHIFT		20
 #define SH_UAC_MASK		(_TIF_UAC_SIGBUS | _TIF_UAC_NOPRINT)
 
 #define SET_UNALIGN_CTL(task,value)	\
-- 
1.6.0.6

